eng/scripts/test-matrix/test-matrix-functions.ps1

```diff
@@ -0,0 +1,317 @@
+    }
+}
+
+function InitializeMatrix {
+    param(
+        [Array]$parameters,
+        [HashTable]$displayNames,
+        [System.Collections.ArrayList]$permutations,
+        [System.Collections.Specialized.OrderedDictionary]$permutation = @{}
+    )
+
+    if (!$parameters) {
+        $entry = CreateMatrixEntry $permutation $displayNames
+        $permutations.Add($entry) | Out-Null
+        return
+    }
+
+    $head, $tail = $parameters
+    foreach ($value in $head.value) {
+        $newPermutation = [ordered]@{}
+        foreach ($element in $permutation.GetEnumerator()) {
+            $newPermutation[$element.Name] = $element.Value
+        }
+        $newPermutation[$head.name] = $value
+        InitializeMatrix $tail $displayNames $permutations $newPermutation
+    }
+}
+
+function GetMatrixDimensions([HashTable]$parameters) {
+    $dimensions = @()
+    foreach ($param in $parameters.GetEnumerator()) {
+        $dimensions += $param.Value.Length
+    }
+
+    return $dimensions | Sort-Object -Descending
+}
+
+function SetNdMatrixElement {
+    param(
+        $element,
+        [ValidateNotNullOrEmpty()]
+        [Array]$idx,
+        [ValidateNotNullOrEmpty()]
+        [Array]$matrix,
+        [ValidateNotNullOrEmpty()]
+        [Array]$dimensions
+    )
+
+    if ($idx.Length -ne $dimensions.Length) {
+        throw "Matrix index query $($idx.Length) must be the same length as its dimensions $($dimensions.Length)"
+    }
+
+    $arrayIndex = GetNdMatrixArrayIndex $idx $dimensions
+    $matrix[$arrayIndex] = $element
+}
+
+function GetNdMatrixArrayIndex {
+    param(
+        [ValidateNotNullOrEmpty()]
+        [Array]$idx,
+        [ValidateNotNullOrEmpty()]
+        [Array]$dimensions
+    )
+
+    if ($idx.Length -ne $dimensions.Length) {
+        throw "Matrix index query length ($($idx.Length)) must be the same as dimension length ($($dimensions.Length))"
+    }
+
+    $stride = 1
+    # Commented out does lookup with wrap handling
+    # $index = $idx[$idx.Length-1] % $dimensions[$idx.Length-1]
+    $index = $idx[$idx.Length-1]
+
+    for ($i = $dimensions.Length-1; $i -ge 1; $i--) {
+        $stride *= $dimensions[$i]
+        # Commented out does lookup with wrap handling
+        # $index += ($idx[$i-1] % $dimensions[$i-1]) * $stride
+        $index += $idx[$i-1] * $stride
+    }
+
+    return $index
+}
+
+function GetNdMatrixElement {
+    param(
+        [ValidateNotNullOrEmpty()]
+        [Array]$idx,
+        [ValidateNotNullOrEmpty()]
+        [Array]$matrix,
+        [ValidateNotNullOrEmpty()]
+        [Array]$dimensions
+    )
+
+    $arrayIndex = GetNdMatrixArrayIndex $idx $dimensions
+    return $matrix[$arrayIndex]
+}
+
+function GetNdMatrixIndex {
+    param(
+        [int]$index,
+        [ValidateNotNullOrEmpty()]
+        [Array]$dimensions
+    )
+
+    $matrixIndex = @()
+    $stride = 1
+
+    for ($i = $dimensions.Length-1; $i -ge 1; $i--) {
+        $stride *= $dimensions[$i]
```

Comment:
Wow `*=` is not something I see often.

Reply:
[This video](https://www.youtube.com/watch?v=DfK83xEtJ_k) was the most helpful for wrapping my head around how to build the indexing algorithm/visualize the data structure. All the other ones are crap :)